### PR TITLE
add scroll buttons that appear in the production palette when icons exceed available space

### LINKED ISSUE #7159: Add a scroll function to the sidebar to avoid clipping on low-res screens.
Hi again guys. Recently a friend on sleipnirstuff noticed after capturing tech from tree factions in my Tiberian Origins mod that you can't build units from the 3rd or 4th side because they are too low on the sidebar and they do not show up. This occurs on
 machines with a resolution of 1366x768 and lower. I had no such problems on my 1920x1080 monitor, but I was able to reproduce the problem on my Asus T100TA. Further clipping occured when I tested the game on a netbook with 1024x600 resolution.

I think this would be fixed if you added a scroll function to the sidebar to accommodate devices with low-resolution screens. 

openra 2014-12-20 21-22-47-95

openra 2014-12-20 21-22-54-25

### PR DESCRIPTION
**Pull request by sinaptik**
Closes #7159

### CODE REVIEW

File: OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs [hunk 1]
```diff
@@ -87,6 +90,35 @@ public ProductionPaletteWidget(OrderManager orderManager, World world, WorldRend
 			clock = new Animation(world, "clock");
 		}
 
+		public void ScrollDown()
+		{
+			var totalRows = AllBuildables.Count() / Columns;
```
Comment (pchote) [inline OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs]: You can skip the check below by changing this to `var totalRows = (AllBuildables.Count() + Columns - 1) / Columns;`

File: OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs [hunk 2]
```diff
@@ -87,6 +90,35 @@ public ProductionPaletteWidget(OrderManager orderManager, World world, WorldRend
...
+				IconRowOffset++;
+		}
+
+		public void ScrollUp()
+		{
+			if (IconRowOffset > 0)
+				IconRowOffset--;
+		}
+
+		public IEnumerable<ActorInfo> AllBuildables
```
Comment (pchote) [inline OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs]: We don't really want to evaluate `AllBuildables.Count()` everywhere (its not a good idea performance-wise), so you should cache both `IconCount` and `DisplayedIconCount` in the `Tick` method.

File: OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs [hunk 3]
```diff
@@ -203,26 +235,25 @@ public void RefreshIcons()
...
-			var oldIconCount = IconCount;
-			IconCount = 0;
+			var oldIconCount = DisplayedIconCount;
+			DisplayedIconCount = 0;
 
 			var ks = Game.Settings.Keys;
 			var rb = RenderBounds;
-			foreach (var item in allBuildables)
+
+			foreach (var item in AllBuildables.Skip(IconRowOffset * Columns).Take(MaxRows * Columns))
```
Comment (pchote) [inline OpenRA.Mods.RA/Widgets/ProductionPaletteWidget.cs]: This causes hotkeys to only be assigned for visible actors, and changes the hotkey for a specific actor when it is scrolled.

**Comment by sinaptik** [comment]
Ugh, not having luck with rebasing. I *think* it's done?

**Comment by sinaptik** [comment]
I *think* the rebase worked. I hope I didn't break anything...

File: OpenRA.Mods.RA/Widgets/Logic/ClassicProductionLogic.cs [hunk 4]
```diff
@@ -147,6 +147,43 @@ public ClassicProductionLogic(Widget widget, OrderManager orderManager, World wo
...
+			if (scrollUp != null)
+			{
+				scrollUp.OnClick = palette.ScrollUp;
+				scrollUp.IsVisible = () => palette.TotalIconCount > (palette.MaxIconRows * palette.Columns);
+			}
+
+			SetMaximumVisibleRows(palette);
+		}
+
+		private static void SetMaximumVisibleRows(ProductionPaletteWidget productionPalette)
```
Comment (Mailaender) [inline OpenRA.Mods.RA/Widgets/Logic/ClassicProductionLogic.cs]: We leave the explicit `private` out by convention. https://github.com/OpenRA/OpenRA/wiki/Coding-Standard

**Comment by Mailaender** [comment]
Needs another rebase.

**Comment by pchote** [comment]
It is sad to see the momentum die out on this pr, as it is very close to being mergeable.  The blame lies squarely on the reviewers (myself included) who are only looking at the most recent two or three PRs and seem to be ignoring everything else.  Unfortunately I don't have time to do general reviewing at this point, so I'm limiting myself to commenting on things that might break things in subtle ways if merged.

Back to this pr: this looks good overall, but there are a couple of minor usability nits remaining.  When running with 600px height the scroll buttons will overflow the panel.  This also ties in with my earlier request to reduce the height of the scroll buttons to help differentiate them.  I made a start on fixing this with 19bec1f75b49f49e8445d9d6f43a6b910c9316c4, but you will need to make the same change to the soviet buttons before this can be merged.  You should also add `IsDisabled` bindings that disable (but not hide) the buttons if the panel is scrolled to the top or bottom.

**Comment by sinaptik** [comment]
Rebased, squashed and requested changes made!

**Comment by obrakmann** [comment]
Found a couple of bugs:

1. with the Build Everything cheat enabled, if you scroll completely down on the support buildings tab and then select the main buildings tab, this happens:
screenshot from 2015-01-22 22 44 21

2. The Scroll Down button doesn't get disabled when you have already scrolled all the way down. The up button however gets correctly disabled when you've scrolled up completely. [This condition](https://github.com/OpenRA/OpenRA/pull/7227/files#diff-8a0c074ae38fba6d0c50b243d2773c4bR159) seems to not work as expected.

Looks good otherwise!

**Comment by sinaptik** [comment]
Nice bug catching obrakmann! I've fixed the scolldown disabled bug, and the only straightforward fix for  the switching tab bug was to reset the rows to the top every time a production tab button is pressed. Otherwise a more complex fix would be to remember the position in each production tab.

**Comment by obrakmann** [comment]
Looks good to me now, :+1: 

Will be nice to see this used in d2k, that needs it the most.

**Comment by obrakmann** [comment]
Thanks!

[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/7e7fe57f0039a4244cc162a09c2d3a123bcc9a93%5E...7e7fe57f0039a4244cc162a09c2d3a123bcc9a93)

**Pull request merged**